libnmstate/state.py

```diff
@@ -236,6 +385,49 @@ def _assert_interfaces_included_in(self, current_state):
                 format_desired_current_state_diff(self.interfaces,
                                                   current_state.interfaces))
 
+    def _clean_routes(self):
+        """
+        Remove routes for down/absent interface
+        Remove routes for non-exit interface
+        Remove routes when IPv4/IPv6 down.
+        """
+        for iface_name, iface_state in six.viewitems(self.interfaces):
+            if iface_state[Interface.STATE] in (InterfaceState.DOWN,
+                                                InterfaceState.ABSENT):
+                if self.iface_routes.get(iface_name, []):
+                    logging.info(
+                        "Removing routes next hop to interface {} "
```

Comment:
Best practice with logging is not to pre-format the string but use `logging.info('%s says %s', the_cow, mooh)` because then the string formatting will only happen if the log message is actually logged. Might not matter here but can improve performance if there are a lot of debug messages that are normally not sent in high performance applications.